utils/fix_xc7_carry.py

```diff
@@ -0,0 +1,595 @@
+                                              |                  |
+                                              | S1           O2  |
+                                              |      CARRY4      |
+                                              | DI0 (chained)    |  +-----------------+
+                                              |                  |  |                 |
+                                              | S0           CO1 +->+ CARRY_CO_DIRECT |
+                                              |                  |  |                 |
+                                              | CYINIT           |  +-----------------+
+                                              |                  |
+                         +-----------------+  |              O1  |
+                         |                 |  |                  |
+                      +->+ CARRY_COUT_PLUG +->+ CI               |  +-----------------+
+                      |  |                 |  |                  |  |                 |
+                      |  +-----------------+  |              CO0 +->+ CARRY_CO_DIRECT |
+                      |                       |                  |  |                 |
+                      |                       |                  |  +-----------------+
+                      +-------------------+   |                  |
+                                          |   |              O0  |
++------------------+  +-----------------+ |   |                  |
+|                  |  |                 | |   +------------------+
+|              CO3 +->+ CARRY_CO_DIRECT +-+
+|                  |  |                 |
+| DI3              |  +-----------------+
+|                  |
+| S3           O3  |
+|                  |
+| DI2              |  +-----------------+
+|                  |  |                 |
+| S2           CO2 +->+ CARRY_CO_DIRECT |
+|                  |  |                 |
+| DI1              |  +-----------------+
+|                  |
+| S1           O2  |
+|       CARRY4     |
+| DI0   (root)     |  +-----------------+
+|                  |  |                 |
+| S0           CO1 +->+ CARRY_CO_DIRECT |
+|                  |  |                 |
+| CYINIT           |  +-----------------+
+|                  |
+|              O1  |
+|                  |
+| CI               |  +-----------------+
+|                  |  |                 |
+|              CO0 +->+ CARRY_CO_DIRECT |
+|                  |  |                 |
+|                  |  +-----------------+
+|                  |
+|              O0  |
+|                  |
++------------------+
+
+
+So there are five cases the script has to handle:
+
+ - No congestion is present between O and CO ->
+
+    Do nothing.
+
+ - Congestion is present on rows 0-2 and row above is in use ->
+
+    Change CARRY_CO_DIRECT to CARRY_CO_LUT.
+
+    Routing and LUT delays are incurred in this case.
+
+ - Congestion is present on rows 0-2 and row above is not in use ->
```

Comment:
You talk about rows here, but there really isn't any "row" markings on the above diagram....

Reply:
Fixed